File: sdk/python/symbolchain/core/TransactionDescriptorProcessor.py
class TransactionDescriptorProcessor:
    """Processes and looks up transaction descriptor properties."""

    def __init__(self, transaction_descriptor, type_parsing_rules, type_converter=None):
        """Creates a transaction descriptor processor."""
        self.transaction_descriptor = transaction_descriptor
        self.type_parsing_rules = type_parsing_rules or {}
        self.type_converter = type_converter or (lambda x: x)
        self.type_hints = {}

    def _lookup_value_and_apply_type_hints(self, key):
        if key not in self.transaction_descriptor:
            raise ValueError('transaction descriptor does not have attribute {}'.format(key))

        value = self.transaction_descriptor[key]

        type_hint = self.type_hints.get(key)
        if type_hint in self.type_parsing_rules:
            value = self.type_parsing_rules[type_hint](value)

        return value

    def lookup_value(self, key):
        """Looks up the value for key."""
        value = self._lookup_value_and_apply_type_hints(key)
        if not any(isinstance(value, type_class) for type_class in [str, bytes, tuple]) and hasattr(value, '__iter__'):
            return [self.type_converter(item) for item in value]

        return self.type_converter(value)

    def copy_to(self, transaction, ignore_keys=None):
        """Copies all descriptor information to a transaction."""
        for key in self.transaction_descriptor.keys():
            if ignore_keys and key in ignore_keys:
                continue

            if not hasattr(transaction, key):
                raise ValueError('transaction does not have attribute {}'.format(key))

            value = self.lookup_value(key)
            if isinstance(value, list):
                getattr(transaction, key).extend(value)
            else:
                setattr(transaction, key, value)

    def set_type_hints(self, type_hints):
        """Sets type hints."""
        self.type_hints = type_hints or {}

File: sdk/javascript/src/TransactionDescriptorProcessor.js
/**
 * Processes and looks up transaction descriptor properties.
 * @note This class is not intended to be used directly.
 */
export default class TransactionDescriptorProcessor {
	/**
	 * Creates a transaction descriptor processor.
	 * @param {object} transactionDescriptor Transaction descriptor.
	 * @param {Map<string, function>} typeParsingRules Type-dependent parsing rules.
	 * @param {function|undefined} typeConverter Converts a generated type to an sdk type (optional).
	 */
	constructor(transactionDescriptor, typeParsingRules, typeConverter = undefined) {
		/**
		 * @private
		 */
		this._transactionDescriptor = transactionDescriptor;

		/**
		 * @private
		 */
		this._typeParsingRules = typeParsingRules;

		/**
		 * Tries to coerce a value to a more appropriate type.
		 * @param {object} value Original value.
		 * @returns {object} Type converted value.
		 * @private
		 */
		this._typeConverter = typeConverter || (value => value);

		/**
		 * @private
		 */
		this._typeHints = {};
	}

	/**
	 * Looks up value and applies type hints.
	 * @param {string} key Key for which to retrieve value.
	 * @returns {object} Value corresponding to key.
	 * @private
	 */
	_lookupValueAndApplyTypeHints(key) {
		if (undefined === this._transactionDescriptor[key])
			throw RangeError(`transaction descriptor does not have attribute ${key}`);

		let value = this._transactionDescriptor[key];
		const typeHint = this._typeHints[key];
		const rule = this._typeParsingRules.get(typeHint);
		if (rule)
			value = rule(value);

		return value;
	}

	/**
	 * Looks up the value for key.
	 * @param {string} key Key for which to retrieve value.
	 * @returns {object} Value corresponding to key.
	 */
	lookupValue(key) {
		const value = this._lookupValueAndApplyTypeHints(key);
		return Array.isArray(value)
			? value.map(item => this._typeConverter(item))
			: this._typeConverter(value);
	}

	/**
	 * Copies all descriptor information to a transaction.
	 * @param {object} transaction Transaction to which to copy keys.
	 * @param {Array<string>|undefined} ignoreKeys Keys of descriptor values not to copy (optional).
	 */
	copyTo(transaction, ignoreKeys = undefined) {
		Object.getOwnPropertyNames(this._transactionDescriptor).forEach(key => {
			if (ignoreKeys && -1 !== ignoreKeys.indexOf(key))
				return;

			if (key.endsWith('Computed'))
				throw RangeError(`cannot explicitly set computed field ${key}`);

			if (undefined === transaction[key])
				throw RangeError(`transaction does not have attribute ${key}`);

			const value = this.lookupValue(key);
			transaction[key] = value;
		});
	}

	/**
	 * Sets type hints.
	 * @param {TypeHintsMap|undefined} typeHints New type hints. // eslint-disable-line valid-jsdoc
	 */
	setTypeHints(typeHints) {
		this._typeHints = typeHints || {};
	}
}

// region type declarations

/**
 * Type hints map.
 * @class
 * @typedef {{[key: string]: string}} TypeHintsMap
 */

// endregion
